### PR TITLE
MSSQL: Fix cannot write binary field in BCP mode

### LINKED ISSUE #3040: MSSQLSpatial: Cannot write binary field when MSSQLSPATIAL_USE_BCP = TRUE
I'm expecting a binary field to write from an OpenFileGDB source into MSSQLSpatial. However, it returns the following error for every feature in the source:
`ERROR 1: INSERT command for new feature failed. [37000][Microsoft][ODBC SQL Server Driver][SQL Server]Incorrect syntax near '0xffd8ffe...`

This happens when I run the command within GDAL 3.1.3 (released 2020/09/01) on Windows 10 obtained from GIS Internals (https://www.gisinternals.com/query.html?content=filelist&file=release-1911-x64-gdal-3-1-3-mapserver-7-6-1.zip):
`ogr2ogr -f "MSSQLSpatial" "MSSQL:server=.;database=MyDB;Uid=sa;Pwd=password;" source.gdb.zip -skipfailures`

I have reproduced the problem on GDAL 2.3.3 (released 2018/12/14) also. I have run MSSQL Profiler to watch the commands generated and noticed that they are each missing the closing ) on the overall INSERT statement:. 

`'SET IDENTITY_INSERT [dbo].[inspections__attach] ON;INSERT INTO [dbo].[inspections__attach] ([attachmentid], [globalid], [rel_globalid], [content_type], [att_name], [data_size], [data]) VALUES (16, @P1, @P2, @P3, @P4, 318830, 0xffd8ffe...d454e308d8',N'@P1 nvarchar(39),@P2 nvarchar(39),@P3 nvarchar(11),@P4 nvarchar(12)',N'{B397675D-739A-4183-97FA-012E889320A6}',N'{D8D13DD1-92CC-4A1A-B7CF-FC171EEDC59F}',N'image/jpeg',N'Photo 2.jpg'`

Please note that same OGR2OGR statement works fine when there are no binary fields involved in the source database. Weird.

I am unfamiliar with the GDAL source, but this looks to be where the closing bracket gets included into the statements, but for some reason it is not being included on these INSERTS - https://github.com/OSGeo/gdal/blob/b51dc28ab7da7c84c1cac3477560dbc9ba9f5c9e/gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlspatialtablelayer.cpp#L2350

### PR DESCRIPTION
**Pull request by elpaso**
Fix #3040

### CODE REVIEW

File: ogr/ogrsf_frmts/mssqlspatial/ogrmssqlspatialtablelayer.cpp [hunk 1]
```diff
@@ -2217,9 +2217,6 @@ OGRErr OGRMSSQLSpatialTableLayer::CreateFeatureBCP(OGRFeature *poFeature)
                         {
                         }
                     }
-                    if (Failed2(bcp_moretext(hDBCBCP, 0, nullptr)))
```
Comment (rouault) [inline ogr/ogrsf_frmts/mssqlspatial/ogrmssqlspatialtablelayer.cpp:2220]: I'm not familiar at all with bcp_moretext(), but removing it here seems to break the "symmetry" with the OFTString and geometry cases above. Any reason why removing it is needed here, and not in those cases ?
Perhaps add in your test case writing a feature with a binary field that is set, but of size 0, to see if there isn't something weird happening (and perhaps also adding another dynamic field, let's say of type String after the binary field, to check that there isn't any mismatch in the bcp batch)

Comment (elpaso) [reply ogr/ogrsf_frmts/mssqlspatial/ogrmssqlspatialtablelayer.cpp:2220]: done, and you were right: it must be called when size is 0.

File: autotest/ogr/ogr_mssqlspatial.py [hunk 2]
```diff
@@ -609,17 +609,21 @@ def test_binary_field_bcp():
...
     f = ogr.Feature(src_lyr.GetLayerDefn())
     f.SetGeometry(ogr.CreateGeometryFromWkt("POINT(1 2)"))
     f.SetFieldBinaryFromHexString("binfield", "007FFF007FFF")
+    f["strfield"] = "some text"
     src_lyr.CreateFeature(f)
     f = ogr.Feature(src_lyr.GetLayerDefn())
     f.SetGeometry(ogr.CreateGeometryFromWkt("POINT(2 3)"))
-    f.SetFieldBinaryFromHexString("binfield", "7FFF007FFF00")
+    f.SetFieldBinaryFromHexString("binfield", "")
+    f["strfield"] = ""
```
Comment (rouault) [inline autotest/ogr/ogr_mssqlspatial.py:620]: I would actually put content here in strfield, to make sure that a 0-size binfield doesn't mess up with non-empty dynamic field content after it

Comment (elpaso) [reply autotest/ogr/ogr_mssqlspatial.py]: done

File: autotest/ogr/ogr_mssqlspatial.py [hunk 3]
```diff
@@ -600,6 +600,67 @@ def test_ogr_mssqlspatial_bulk_insert():
 #
 
 
+@pytest.mark.require_driver("GPKG")
+def test_binary_field_bcp():
+    """Test for issue GH #3040 MSSQLSpatial: Cannot write binary field with MSSQLSPATIAL_USE_BCP = TRUE"""
+
```
Comment (rouault) [inline autotest/ogr/ogr_mssqlspatial.py:606]: to skip the test when the driver is compiled, but the connection string not defined: (ideally we'd have a module-level pytest.skip.if() to do that, like in other similar drivers that require specific parameters to be passed)

```suggestion

    if gdaltest.mssqlspatial_ds is None:
        pytest.skip()

```

And you'd need to solve conflict with master now that your other MSSQL PR has been merged

**Comment by rouault** [comment]
The backport to `release/3.7` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply d513948251... MSSQL: Fix cannot write binary field in BCP mode
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging autotest/ogr/ogr_mssqlspatial.py
CONFLICT (content): Merge conflict in autotest/ogr/ogr_mssqlspatial.py

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-release/3.7 release/3.7
# Navigate to the new working tree
cd .worktrees/backport-release/3.7
# Create a new branch
git switch --create backport-7841-to-release/3.7
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick d513948251edc7c3326919958dfa138865906c1d,bfae23dd62a7fe8cd33423c55ee37be001decd48,787f02234213edfa69b86e829ada449584bb1a0e,6ce59083bed3c5e1fdd19b466bb24c85f90d155c,85d752a1b9dbd231aa1d57484bc4b195f775cc30
# Push it to GitHub
git push --set-upstream origin backport-7841-to-release/3.7
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-release/3.7
```
Then, create a pull request where the `base` branch is `release/3.7` and the `compare`/`head` branch is `backport-7841-to-release/3.7`.

**Comment by rouault** [comment]
backport in https://github.com/OSGeo/gdal/pull/7843

**Pull request merged**